### PR TITLE
[FIX] account_edi_ubl_cii: select the format based on the partner

### PR DESCRIPTION
**Pull request by JulienVR**
Add a selection field on the partner to choose the EDI format. Merge the `cii_` and `ubl_` fields.

Add an `endpoint_value` and `eas_code` on the partner. These values are read when generating the Bis 3 (or one of its derivatives) to fill the `EndpointID` and the `@schemeID` in the xml.

We can remove `l10n_nl_oin` and `l10n_nl_kvk` which are only used by electronic invoicing, and replace every usage of them by the new fields `endpoint_value` and `eas_code` (since for NL partners, they should necessarily represent the KVK or OIN number: eas code '0106' or '0190').

We can also remove the `l10n_lu_peppol_id` module which was only there to allow an arbitrary number to be set on the `EndpointID` (for instance, for public administration that doesn't have a vat number).

task-3232845

upgrade: https://github.com/odoo/upgrade/pull/4445

### CODE REVIEW

File: addons/account_edi_ubl_cii/models/account_move.py [hunk 1]
```diff
@@ -5,27 +5,15 @@
...
         comodel_name='ir.attachment',
-        string="UBL Attachment",
-        compute=lambda self: self._compute_linked_attachment_id('ubl_xml_id', 'ubl_xml_file'),
-        depends=['ubl_xml_file']
+        string="Attachment",
+        compute=lambda self: self._compute_linked_attachment_id('xml_id', 'xml_file'),
+        depends=['xml_file']
     )
-    ubl_xml_file = fields.Binary(
+    xml_file = fields.Binary(
```
Comment (tdetry) [inline addons/account_edi_ubl_cii/models/account_move.py:14]: make explicit ubl/cii

File: addons/account_edi_ubl_cii/models/account_move_send.py [hunk 2]
```diff
@@ -16,59 +16,43 @@
...
-    enable_cii_xml = fields.Boolean(compute='_compute_send_mail_extra_fields')
-    checkbox_cii_xml = fields.Boolean(
-        string="CII Billing",
-        compute='_compute_checkbox_cii_xml',
-        store=True,
-        readonly=False,
-    )
+    enable_xml = fields.Boolean(compute='_compute_send_mail_extra_fields')
+    checkbox_label = fields.Char(compute='_compute_checkbox_label')  # label for the checkbox_xml field
+    checkbox_xml = fields.Boolean(compute='_compute_checkbox_xml', store=True, readonly=False)
```
Comment (tdetry) [inline addons/account_edi_ubl_cii/models/account_move_send.py:21]: make explicit ubl/cii

File: addons/account_edi_ubl_cii/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models, fields, api, _
+from odoo.addons.account_edi_ubl_cii.models.account_edi_common import EAS_MAPPING
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    edi_format = fields.Selection(
```
Comment (tdetry) [inline addons/account_edi_ubl_cii/models/res_partner.py:11]: make explicit ubl/cii

File: addons/account_edi_ubl_cii/models/account_move_send.py [hunk 4]
```diff
@@ -174,11 +154,12 @@ def _postprocess_document(self, invoice, prepared_data):
...
         writer.cloneReaderDocumentRoot(reader)
 
-        writer.addAttachment(cii_xml['name'], cii_xml['raw'], subtype='text/xml')
+        writer.addAttachment('factur-x.xml', xml_facturx, subtype='text/xml')
 
         # PDF-A.
-        options = prepared_data['cii_xml_options']['options']
-        if options.get('facturx_pdfa') and not writer.is_pdfa:
+        if 'xml_options' in prepared_data \
+                and prepared_data['xml_options']['options'].get('facturx_pdfa') \
```
Comment (tdetry) [inline addons/account_edi_ubl_cii/models/account_move_send.py:161]: wondering if we should keep this option; facturx_pdfa. or we can just check if 'facturx'

**Review by tdetry — COMMENTED** [review]
the change was needed :-)  thanks for taking care of it!

File: addons/account_edi_ubl_cii/models/account_move.py [hunk 5]
```diff
@@ -5,27 +5,15 @@
...
-    cii_xml_id = fields.Many2one(
-        comodel_name='ir.attachment',
-        string="CII Attachment",
-        compute=lambda self: self._compute_linked_attachment_id('cii_xml_id', 'cii_xml_file'),
-        depends=['cii_xml_file']
-    )
-    cii_xml_file = fields.Binary(
-        comodel_name='ir.attachment',
-        string="CII File",
+        string="File",
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/account_move.py:16]:
```suggestion
        string="UBL/CII File",
```

File: addons/account_edi_ubl_cii/models/account_move_send.py [hunk 6]
```diff
@@ -16,59 +16,43 @@
...
     # COMPUTE METHODS
     # -------------------------------------------------------------------------
 
+    @api.depends('move_ids')
+    def _compute_checkbox_ubl_cii_label(self):
+        for wizard in self:
+            code_to_label = dict(wizard.move_ids.partner_id._fields['ubl_cii_format'].selection)
+            codes = wizard.move_ids.partner_id.mapped('ubl_cii_format')
+            if any(codes):
+                wizard.checkbox_ubl_cii_label = ", ".join([code_to_label[c] for c in codes])
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/account_move_send.py:37]:
```suggestion
                wizard.checkbox_ubl_cii_label = ", ".join(code_to_label[c] for c in codes)
```

File: addons/account_edi_ubl_cii/models/account_move_send.py [hunk 7]
```diff
@@ -16,59 +16,43 @@
...
+            else:
+                wizard.checkbox_ubl_cii_label = False
+
     def _compute_send_mail_extra_fields(self):
         # EXTENDS 'account'
         super()._compute_send_mail_extra_fields()
         for wizard in self:
-            wizard.enable_ubl_xml = any([self._get_default_enable_ubl_xml(m) for m in wizard.move_ids])
-            wizard.enable_cii_xml = any([self._get_default_enable_cii_xml(m) for m in wizard.move_ids])
+            wizard.enable_ubl_cii_xml = any([self._get_default_enable_ubl_cii_xml(m) for m in wizard.move_ids])
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/account_move_send.py:45]:
```suggestion
            wizard.enable_ubl_cii_xml = any(self._get_default_enable_ubl_cii_xml(m) for m in wizard.move_ids)
```

File: addons/account_edi_ubl_cii/models/account_move_send.py [hunk 8]
```diff
@@ -101,50 +85,37 @@ def _get_placeholder_mail_attachments_data(self, move):
...
-        xmls_to_prepare = []
-
-        # CII
-        cii_xml = self.checkbox_cii_xml and self._get_default_enable_cii_xml(invoice)
-        if cii_xml:
-            builder, options = invoice.journal_id._get_cii_builder()
-        else:
-            # Default XML acting as the default EDI for ODOO hidden from the users (facturx).
-            builder, options = self.env['account.edi.xml.cii'], {'skip_errors': True}
+        invoice.ensure_one()
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/account_move_send.py:88]: it shouldn't be necessary

File: addons/account_edi_ubl_cii/models/account_move_send.py [hunk 9]
```diff
@@ -101,50 +85,37 @@ def _get_placeholder_mail_attachments_data(self, move):
...
-        for edi_type, builder, options in xmls_to_prepare:
             xml_content, errors = builder._export_invoice(invoice)
             filename = builder._export_invoice_filename(invoice)
 
             # Failed.
             if errors and not options.get('skip_errors'):
                 return {
                     'error': "".join([
-                        _("Errors occur while creating the EDI document (format: %s):", builder._description),
+                        _("Errors occured while creating the EDI document (format: %s):", builder._description),
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/account_move_send.py:102]: change the corresponding translation then

File: addons/account_edi_ubl_cii/models/account_move_send.py [hunk 10]
```diff
@@ -101,50 +85,37 @@ def _get_placeholder_mail_attachments_data(self, move):
...
                         "\n",
                         "<p><li>" + "</li><li>".join(errors) + "</li></p>" if self.mode == 'invoice_multi' \
                             else "\n".join(errors)
                     ]),
                 }
 
-            results[f'{edi_type}_xml_attachment_values'] = {
+            results['xml_attachment_values'] = {
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/account_move_send.py:109]:
```suggestion
            results['ubl_cii_xml_attachment_values'] = {
```

File: addons/account_edi_ubl_cii/models/account_move_send.py [hunk 11]
```diff
@@ -101,50 +85,37 @@ def _get_placeholder_mail_attachments_data(self, move):
...
                 'name': filename,
                 'raw': xml_content,
                 'mimetype': 'application/xml',
                 'res_model': invoice._name,
                 'res_id': invoice.id,
-                'res_field': f'{edi_type}_xml_file', # Binary field
+                'res_field': 'ubl_cii_xml_file', # Binary field
             }
-            results[f'{edi_type}_xml_options'] = {
+            results['xml_options'] = {
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/account_move_send.py:117]:
```suggestion
            results['ubl_cii_xml_options'] = {
```

File: addons/account_edi_ubl_cii/models/account_move_send.py [hunk 12]
```diff
@@ -155,15 +126,24 @@ def _postprocess_document(self, invoice, prepared_data):
         # EXTENDS 'account'
         super()._postprocess_document(invoice, prepared_data)
 
-        cii_xml = prepared_data.get('cii_xml_attachment_values')
-        ubl_xml = prepared_data.get('ubl_xml_attachment_values')
-
         # Add PDF to XML
-        if ubl_xml:
+        if 'xml_options' in prepared_data and prepared_data['xml_options']['ubl_cii_format'] != 'facturx':
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/account_move_send.py:130]: why not 
```py
if prepared_data.get('ubl_cii_xml_attachment_values'):
    self._postprocess_invoice_ubl_xml(invoice, prepared_data)
else:
    # facturx pdf-a stuff?
```

File: addons/account_edi_ubl_cii/models/res_partner.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+            elif partner.country_code in EAS_MAPPING:
+                partner.ubl_cii_format = 'ubl_bis3'
+            else:
+                partner.ubl_cii_format = partner.ubl_cii_format
+
+    @api.depends('country_code', 'vat')
+    def _compute_peppol_endpoint(self):
+        for partner in self:
+            if partner.ubl_cii_format != 'facturx' \
+                    and partner.country_code in EAS_MAPPING:
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/res_partner.py:144]:
```suggestion
            if partner.ubl_cii_format != 'facturx' and partner.country_code in EAS_MAPPING:
```

File: addons/account_edi_ubl_cii/models/res_partner.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+                eas_to_field = EAS_MAPPING.get(partner.country_code)
+                # Try to set both the peppol_eas and the peppol_endpoint
+                for _eas, field_name in eas_to_field.items():
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/res_partner.py:147]:
```suggestion
                for field_name in eas_to_field.values():
```

File: addons/account_edi_ubl_cii/models/res_partner.py [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+                if not partner.peppol_endpoint:
+                    partner.peppol_endpoint = partner.peppol_endpoint
+            else:
+                partner.peppol_endpoint = partner.peppol_endpoint
+
+    @api.depends('country_code', 'vat')
+    def _compute_peppol_eas(self):
+        for partner in self:
+            if partner.ubl_cii_format != 'facturx' \
+                    and partner.country_code in EAS_MAPPING:
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/res_partner.py:160]:
```suggestion
            if partner.ubl_cii_format != 'facturx' and partner.country_code in EAS_MAPPING:
```

File: addons/account_edi_ubl_cii/models/res_partner.py [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+                eas_to_field = EAS_MAPPING.get(partner.country_code)
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/res_partner.py:161]:
```suggestion
                eas_to_field = EAS_MAPPING[partner.country_code]
```

File: addons/account_edi_ubl_cii/models/account_move_send.py (hunk 4 above)
Comment (JulienVR) [reply addons/account_edi_ubl_cii/models/account_move_send.py:161]: good catch !

File: addons/account_edi_ubl_cii/models/res_partner.py (hunk 13 above)
Comment (JulienVR) [reply addons/account_edi_ubl_cii/models/res_partner.py:144]: oups

**Comment by smetl** [comment]
@robodoo r+

**Comment by Xavier-Do** [comment]
(should be green in a few minutes)

**Pull request closed without merging**